time_interp/time_interp_external2.F90

```diff
@@ -1094,33 +808,62 @@ subroutine load_record(field, rec, interp, is_in, ie_in, js_in, je_in, window_id
+        mask_in = 0.0_r8_kind
+        where (is_valid(field%src_data(:,:,:,ib), field%valid)) mask_in = 1.0_r8_kind
         if ( field%region_type .NE. NO_REGION ) then
-           if( ANY(mask_in == 0.0) ) then
+           if( ANY(mask_in == 0.0_r8_kind) ) then
               call mpp_error(FATAL, "time_interp_external: mask_in should be all 1 when region_type is not NO_REGION")
            endif
            if( field%region_type == OUTSIDE_REGION) then
               do j = js_region, je_region
                  do i = is_region, ie_region
-                    mask_in(i,j,:) = 0.0
+                    mask_in(i,j,:) = 0.0_r8_kind
                  enddo
               enddo
            else  ! field%region_choice == INSIDE_REGION
               do j = 1, size(mask_in,2)
                  do i = 1, size(mask_in,1)
-                    if( j<js_region .OR. j>je_region .OR. i<is_region .OR. i>ie_region ) mask_in(i,j,:) = 0.0
+                    if( j<js_region .OR. j>je_region .OR. i<is_region .OR. i>ie_region ) mask_in(i,j,:) = 0.0_r8_kind
                  enddo
               enddo
            endif
         endif
-        allocate(mask_out(isw:iew,jsw:jew, size(field%src_data,3)))
-        call horiz_interp(interp,field%src_data(:,:,:,ib),field%data(isw:iew,jsw:jew,:,ib), &
-             mask_in=mask_in, &
-             mask_out=mask_out)
+        !! added for mixed mode. if existing horiz_interp_type was initialized in r4, needs to cast down in order
+        !! to match up with saved values in horiz_interp_type.
+        !! creates some temporary arrays since intent(out) vars can't get passed in diretory
+        if (interp%horizInterpReals4_type%is_allocated) then
+            ! allocate (there may be a better way to do this, had issues with gnu)
+            allocate(hi_tmp_msk_out(isw:iew,jsw:jew, SIZE(field%src_data,3)))
+            allocate(hi_tmp_data(LBOUND(field%data,1):UBOUND(field%data,1), &
+                                 LBOUND(field%data,2):UBOUND(field%data,2), &
+                                 LBOUND(field%data,3):UBOUND(field%data,3), &
+                                 LBOUND(field%data,4):UBOUND(field%data,4)))
+            allocate(hi_tmp_src_data(LBOUND(field%src_data,1):UBOUND(field%src_data,1), &
+                                     LBOUND(field%src_data,2):UBOUND(field%src_data,2), &
+                                     LBOUND(field%src_data,3):UBOUND(field%src_data,3), &
+                                     LBOUND(field%src_data,4):UBOUND(field%src_data,4)))
+            ! assign if needed
+            hi_tmp_data = real(field%data, r4_kind)
+            hi_tmp_src_data = real(field%src_data, r4_kind)
+            ! do interpolation
+            call horiz_interp(interp, hi_tmp_src_data(:,:,:,ib), hi_tmp_data(isw:iew,jsw:jew,:,ib), &
+                              mask_in=real(mask_in,r4_kind), mask_out=hi_tmp_msk_out)
+            ! assign any output
+            field%data = real(hi_tmp_data, r8_kind)
```

Comment:
Why are we casting back to r8 in this case?

Reply:
The `field%data` is read in from the field always in r8_kind with read_data above. In this case a horiz_interp_type was passed in for interpolation in r4_kind so it makes a copy to pass in, then assigns the output.

Reply:
I guess casting ingest data from r8 -> r4 and interpolating using the stored info in the Interp type will most likely have the same amount of error as casting the Interp type data to r8 and performing the interpolation.  I should also add tht though we are storing the data from the file as r8, we don't know what the actual precision of the data in the file itself.  Do we have any inline comments explaining these inconsistencies?